File: RCTTerra/RCTTerra.m
//
//  RCTTerra.m
//  RCTTerra
//
//  Created by Jaafar Rammal on 03/01/2022.
//

#import "React/RCTBridgeModule.h"
@interface RCT_EXTERN_REMAP_MODULE(RNTerraApple, TerraSwiftBridge, NSObject)

// auth
RCT_EXTERN_METHOD(
    auth:
    (NSString *)devID
    apiKey: (NSString *)apiKey
    referenceID: (NSString *)referenceID
    resolve: (RCTPromiseResolveBlock)resolve
    rejecter: (RCTPromiseRejectBlock)reject
)
RCT_EXTERN_METHOD(deauth)


// init
RCT_EXTERN_METHOD(
    initTerra:
    (NSString *)devID
    apiKey: (NSString *)apiKey
    autoFetch: (BOOL *)autoFetch
    resolve: (RCTPromiseResolveBlock)resolve
    rejecter: (RCTPromiseRejectBlock)reject
)

// getters
RCT_EXTERN_METHOD(
    getAthlete: (RCTPromiseResolveBlock)resolve
    rejecter: (RCTPromiseRejectBlock)reject
)
RCT_EXTERN_METHOD(
    getBody: (NSDate *)startDate
    endDate:(NSDate *)endDate
    resolve: (RCTPromiseResolveBlock)resolve
    rejecter: (RCTPromiseRejectBlock)reject
)
RCT_EXTERN_METHOD(
    getDaily: (NSDate *)startDate
    endDate:(NSDate *)endDate
    resolve: (RCTPromiseResolveBlock)resolve
    rejecter: (RCTPromiseRejectBlock)reject
)
RCT_EXTERN_METHOD(
    getSleep: (NSDate *)startDate
    endDate:(NSDate *)endDate
    resolve: (RCTPromiseResolveBlock)resolve
    rejecter: (RCTPromiseRejectBlock)reject
)
RCT_EXTERN_METHOD(
    getActivity: (NSDate *)startDate
    endDate:(NSDate *)endDate
    resolve: (RCTPromiseResolveBlock)resolve
    rejecter: (RCTPromiseRejectBlock)reject
)
RCT_EXTERN_METHOD(
    getNutrition: (NSDate *)startDate
    endDate:(NSDate *)endDate
    resolve: (RCTPromiseResolveBlock)resolve
    rejecter: (RCTPromiseRejectBlock)reject
)

@end

File: RCTTerra/RCTTerra.swift
//
//  RCTTerra.swift
//  RCTTerra
//
//  Created by Jaafar Rammal on 03/01/2022.
//

import Foundation
import TerraSwift

@objc(TerraSwiftBridge)
class TerraSwiftBridge: NSObject {
  
  // some states
  private var count = 0
  private var terraClient: TerraSwift.Terra?
  
  //  require init on main
  @objc
  static func requiresMainQueueSetup() -> Bool {
    return true
  }
  
  // init client
  @objc
  func auth(_ devID: String, apiKey: String, referenceID: String, resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
    guard let terraAuthResponse = TerraSwift.connectTerra(dev_id: devID, xAPIKey: apiKey, referenceId: referenceID) else {
      reject("AUTH", "Auth failed, further debug messages avaialble in Xcode", nil)
      return
    }
    print(terraAuthResponse)
    resolve(["userID": terraAuthResponse.user_id, "referenceID": terraAuthResponse.reference_id, "status": terraAuthResponse.status])
  }
    
  @objc
  func initTerra(_ devID: String, apiKey: String, autoFetch: Bool, resolve: @escaping RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock){
    do {
      terraClient = try TerraSwift.Terra(dev_id: devID, xAPIKey: apiKey, auto: autoFetch){(success: Bool) in resolve(["success": success])}
    } catch {
      reject("Init", "Init failed, further debug messages avaialble in Xcode", nil)
    }
  }
  
  // deauth
  @objc
  func deauth(){
    terraClient = nil
  }
  
  // data functions
  @objc
  func getBody(_ startDate: Date, endDate: Date, resolve: @escaping RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock){
    terraClient?.getBody(startDate: startDate, endDate: endDate){(success: Bool) in resolve(["success": success])}
  }
  @objc
  func getDaily(_ startDate: Date, endDate: Date, resolve: @escaping RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock){
    terraClient?.getDaily(startDate: startDate, endDate: endDate){(success: Bool) in resolve(["success": success])}
  }
  @objc
  func getSleep(_ startDate: Date, endDate: Date, resolve: @escaping RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock){
    terraClient?.getSleep(startDate: startDate, endDate: endDate){(success: Bool) in resolve(["success": success])}
  }
  @objc
  func getActivity(_ startDate: Date, endDate: Date, resolve: @escaping RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock){
    terraClient?.getWorkout(startDate: startDate, endDate: endDate){(success: Bool) in resolve(["success": success])}
  }
  @objc
  func getAthlete(resolve: @escaping RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock){
    terraClient?.getAthlete(){(success: Bool) in resolve(["success": success])}
  }
  
}

